front/app/containers/ProjectsShowPage/timeline/PhaseNavigation.tsx

```diff
@@ -69,18 +74,29 @@ interface Props {
 const PhaseNavigation = memo<Props & WrappedComponentProps>(
   ({ projectId, buttonStyle, className, intl: { formatMessage } }) => {
     const { data: phases } = usePhases(projectId);
+    const { phaseNumber } = useParams();
+    const { data: project } = useProjectById(projectId);
 
-    const [selectedPhase, setSelectedPhase] = useState<
-      IPhaseData | undefined
-    >();
+    const selectedPhase = useMemo(() => {
+      if (!phases) return;
 
-    useEffect(() => {
-      const subscription = selectedPhase$.subscribe((selectedPhase) => {
-        setSelectedPhase(selectedPhase);
-      });
+      // if a phase parameter was provided, and it is valid, we set that as phase.
+      // otherwise, use the most logical phase
+      if (isValidPhase(phaseNumber, phases.data)) {
+        const phaseIndex = Number(phaseNumber) - 1;
+        return phases.data[phaseIndex];
+      }
 
-      return () => subscription.unsubscribe();
-    }, []);
+      return getLatestRelevantPhase(phases.data);
+    }, [phaseNumber, phases]);
+
+    const selectPhase = useCallback(
```

Comment:
Did you add useCallback because there was a performance problem?

(Already giving my thoughts, regardless of the answer: I think in general we should go easier on `useCallback` and `useMemo` in our codebase and only use if it we measure + experience a real problem. It's extra JS (to parse, download, etc.) and makes the code look more complicated.)

Reply:
I think we should talk about this at the CDM. Both you and Iva seem very anti `useMemo` and `useCallback`, while I think that sometimes it's good to use them (although not always). For me the reason to use them are:

1.  When I am refactoring something, and it was already using `useCallback` or `useMemo`, I'd rather not mess with it. That's what is happening here. Same if I am refactoring a class component to a functional component- the class component methods are stable references, so in order to not break stuff by accident I rather keep them stable references. This is particularly an issue with child components using `useEffect`, which can cause infinite loops.
2. Obsolete re-renders causing yanky behavior and a bad user experience. In many parts of our codebase we don't properly test or understand how often components re-render and what components are doing during the rendering lifecycle. This is causing the UI to be yanky which looks bad and unprofessional and makes it annoying to use sometimes. Often this is caused by obsolete re-renders which are in turn caused partly by not memoizing non-primitives where we should memoize them.
3. I don't agree that we should only use if if we measure + experience a real problem. Our light house score is terrible and part of the reason is 'performance death by a thousand cuts', in other words, a lot of small pieces of bad code spread everywhere. Yes, if we want to improve performance we can focus on low hanging fruit first, but at some point you won't get a higher score if you just keep writing non-performant code.

Also, I am not very convinced by the whole 'the code looks more complicated' argument. It's always exactly the same pattern and once you know it you don't think twice about it.